sql/server.go

```diff
@@ -155,13 +156,29 @@ func (p parameters) Arg(i int) (parser.Datum, bool) {
 // exec executes the request. Any error encountered is returned; it is
 // the caller's responsibility to update the response.
 func (s *Server) exec(req driver.Request) (resp driver.Response, err error) {
-
 	// Pick up current session state.
 	// The request user is validated in ServeHTTP. Even in insecure mode,
 	// it is guaranteed not to be empty.
 	planMaker := planner{user: req.GetUser()}
 	defer func() {
 		// Update session state even if an error occurs.
+		if err != nil && planMaker.txn != nil {
+			// If we hit an error and there is a pending transaction, rollback
+			// the transaction before returning. The client does not have to
+			// deal with cleaning up transaction state.
+			if rollbackErr := planMaker.txn.Rollback(); rollbackErr != nil {
+				err = rollbackErr
+			}
+			// reset.
+			planMaker.txn = nil
+		}
+		// Update session state.
+		if planMaker.txn != nil {
+			t := planMaker.txn.GetState()
```

Comment:
Changed the name to ToProto. I'd made a change recently to have Txn.txn exported and then realized it was not worth having Txn.txn changed from under the transaction. Having ToProto return a pointer would expose Txn to the same. I like the return by value.